pkg/chunk/cache/diskcache.go

```diff
@@ -116,25 +160,42 @@ func (d *Diskcache) fetch(key string) ([]byte, bool) {
 
 // StoreChunk puts a chunk into the cache.
 func (d *Diskcache) StoreChunk(ctx context.Context, key string, value []byte) error {
-	d.mtx.Lock()
-	defer d.mtx.Unlock()
-
 	bucket := hash(key) % d.buckets
+
+	mutex := bucket % numMutexes // Get the index of the mutex associated with this bucket
+	d.indexMutexes[mutex].Lock()
+	defer d.indexMutexes[mutex].Unlock()
+
+	if d.index[bucket] == key { // If chunk is already cached return nil
+		return nil
+	}
+
+	if d.index[bucket] == "" {
+		bucketsInitialized.Inc()
+	} else {
+		collisionsTotal.Inc()
+	}
+
 	buf := d.buf[bucket*bucketSize : (bucket+1)*bucketSize]
 
 	n, err := put([]byte(key), buf, 0)
 	if err != nil {
+		d.index[bucket] = ""
 		return err
 	}
```

Comment:
Now that we have the keys in memory, do we need to write them to the mmap'd file as well?

Reply:
I actually tried this and it caused the checksum errors. I've been looking into the root cause and I think I might have found it. Going to test it out today.